internal/uuid-endianness/uuid.go

```diff
@@ -20,65 +22,53 @@ type Uuid struct {
 	Node             [6]byte
 }
 
-// String implements the fmt.Stringer interface
-func (u Uuid) String() string {
-	return fmt.Sprintf("%08x-%04x-%04x-%02x%02x-%x", u.TimeLow, u.TimeMid, u.TimeHiAndVersion, u.ClockSeqHiAndRes, u.ClockSeqLow, u.Node)
-}
-
-func FromString(s string) (Uuid, error) {
-	var uid Uuid
-	u, err := guuid.Parse(s)
+// Redfish API and probably the DHCP Request return the GUID of a machine encoded either
+// in middleEndianBytes
+// or as normal string
+//
+// To detect which one is actually in use, we extract the creation time of the uuid,
+// if this time is somehow reasonable, we return the string of the GUID as we got it,
+// we need to convert it to mixed endian
+// https://en.wikipedia.org/wiki/Universally_unique_identifier#Encoding
+func Convert(s string) (string, error) {
+	u, err := uuid.Parse(s)
 	if err != nil {
-		return uid, err
+		return "", err
+	}
+	if isNotEncoded(u) {
+		return u.String(), nil
 	}
+
 	b, err := u.MarshalBinary()
 	if err != nil {
-		return uid, err
+		return "", err
 	}
-	return FromBytes(b)
-}
-
-// ToMiddleEndian encodes the UUID into a middle-endian UUID.
-//
-// A middle-endian encoded UUID represents a UUID where the first
-// three groups are Little Endian encoded, while the rest of the
-// groups are Big Endian encoded.
-func (u Uuid) ToMiddleEndian() (Uuid, error) {
-	buf, err := u.MiddleEndianBytes()
+	uid, err := fromBytes(b)
 	if err != nil {
-		return Uuid{}, err
+		return "", err
 	}
-
-	return FromBytes(buf)
-}
-
-// BigEndianBytes returns the UUID encoded in Big Endian.
-func (u Uuid) BigEndianBytes() ([]byte, error) {
-	buf := bytes.NewBuffer(make([]byte, 0, UuidSize))
-
-	err := binary.Write(buf, binary.BigEndian, u)
+	buf, err := uid.middleEndianBytes()
 	if err != nil {
-		return nil, err
+		return "", err
 	}
-
-	return buf.Bytes(), nil
+	uid, err = fromBytes(buf)
+	return fmt.Sprintf("%08x-%04x-%04x-%02x%02x-%x", uid.TimeLow, uid.TimeMid, uid.TimeHiAndVersion, uid.ClockSeqHiAndRes, uid.ClockSeqLow, uid.Node), err
 }
 
-// LittleEndianBytes returns the UUID encoded in Little Endian.
-func (u Uuid) LittleEndianBytes() ([]byte, error) {
-	buf := bytes.NewBuffer(make([]byte, 0, UuidSize))
+// thirtyYears is an educated guess for a plausible time stored in the uuid.
+// RFC states that the time is stored in 100s of nanos since 15 Okt 1582 as defined in RFC 4122
+// We check if this time is not more than 30 years apart from now.
+// If the uid returned from the BMC is mixedEndian encoded, the time extracted is usually in the year 4000 or so.
+const thirtyYears = 30 * 365 * 24 * time.Hour
 
-	err := binary.Write(buf, binary.LittleEndian, u)
-	if err != nil {
-		return nil, err
-	}
-
-	return buf.Bytes(), nil
+func isNotEncoded(u uuid.UUID) bool {
+	timeDistance := time.Since(time.Unix(u.Time().UnixTime())).Abs()
+	return timeDistance < thirtyYears
```

Comment:
the impitool implementation additionally checks if the timestamp is not in the future:

```
	if (tm.tm_year > now.tm_year) {
		/* GUID timestamp can't be in future */
```

Reply:
OK, done